metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractValueCodeAction.scala

```diff
@@ -33,55 +33,131 @@ class ExtractValueCodeAction(
     val path = params.getTextDocument().getUri().toAbsolutePath
     val range = params.getRange()
 
-    val applyOpt = trees.findLastEnclosingAt[Term.Apply](
-      path,
-      range.getStart(),
-      appl => appl.args.exists(_.pos.encloses(range))
-    )
-
-    val textEdits = for {
-      apply <- applyOpt
-      stats <- lastEnclosingStatsList(apply)
-      argumentOpt <- apply.args.find { arg => arg.pos.encloses(range) }
-      argument = argumentOpt match {
-        // named paramaeter
-        case Term.Assign(_, rhs) => rhs
-        case other => other
+    val allTrees: Seq[Term] = {
+      def loop(t: Tree): Seq[Term] = {
+        t.children.find(_.pos.encloses(range)) match {
+          case Some(tr: Term) if existsRangeEnclosing(tr, range) =>
+            loop(tr) :+ tr
```

Comment:
If we do `tr :: loop(tr)`, should we also reverse the list after to get the resulting extract options from smallest range?

Reply:
Yeah, sure.

> `:+` on List (List is default Seq implementation) appends alement to the end of collection. Each time whole collection needs to be traversed so it has linear complexity. `::` on the other hand prepends element and has O(1) complexity.

Obviously in this case this doesn't make any difference since here in 99% there will be only one element in that List but it's better to do it because in general it won't be necessary similar case.